threepid_checker/__init__.py

```diff
@@ -29,18 +31,68 @@ def __init__(self, config: ThreepidCheckerConfig, api: ModuleApi):
         self._api = api
         self._config = config
 
+        self._http_client = api.http_client
+
+        self._api.register_password_auth_provider_callbacks(
+            is_3pid_allowed=self.check_if_allowed,
+        )
+
     @staticmethod
     def parse_config(config: Dict[str, Any]) -> ThreepidCheckerConfig:
-        # Parse the module's configuration here.
-        # If there is an issue with the configuration, raise a
-        # synapse.module_api.errors.ConfigError.
-        #
-        # Example:
-        #
-        #     some_option = config.get("some_option")
-        #     if some_option is None:
-        #          raise ConfigError("Missing option 'some_option'")
-        #      if not isinstance(some_option, str):
-        #          raise ConfigError("Config option 'some_option' must be a string")
-        #
-        return ThreepidCheckerConfig()
+        """Checks that required config options are in the correct format and parses them
+        into a ThreepidCheckerConfig.
+
+        Args:
+            config: The raw configuration dict.
+
+        Returns:
+            The parsed configuration.
+        """
+        if "url" not in config:
+            raise ConfigError('"url" is a required configuration parameter')
+
+        if not isinstance(config["url"], str) or not config["url"].startswith("http"):
+            raise ConfigError('"url" needs to be an HTTP(S) URL')
+
+        return ThreepidCheckerConfig(**config)
+
+    async def check_if_allowed(
+        self, medium: str, address: str, registration: bool
+    ) -> bool:
+        """Sends an HTTP(s) request to the configured URL and check if the data it
+        responds with allows the given 3PID to be associated with a local account.
+
+        Note that this function does not check if an error is raised by the HTTP client.
+        The idea is that Synapse will catch that error, log it and fail the user's
+        request, which is what we want anyway.
+
+        Args:
+            medium: The 3PID's medium.
+            address: The 3PID's address.
+            registration: Whether the check is happening while registering a new user.
+
+        Returns:
+            Whether the 3PID can register.
+        """
+        if registration is False and self._config.only_check_at_registration is True:
+            # TODO: test
+            return True
+
+        data = await self._http_client.get_json(
+            self._config.url,
+            {"medium": medium, "address": address},
+        )
```

Comment:
Probably doesn't matter, but this is *not* subject to the URL blacklisting in Synapse. The URL is from our configuration though so I think it is fine!

Reply:
Yeah the URL is supposed to be an internal one (hence `internal-info` in Tchap's spec). Good point though, I'll add a mention about URL blacklisting to the readme.